datasharing - Redshift adapter (#8920)
fishtown-analytics/docs.getdbt.com: website/docs/docs/local/connect-data-platform/redshift-setup.md

| `role`  | None | Optional, user identifier of the current session |
| `autocreate`  | false | Optional, default `False`. Creates user if they do not exist |
| `db_groups`  | ['ANALYSTS'] | Optional. A list of existing database group names that the DbUser joins for the current session |

> | `ra3_node`  | true | Optional, default `False`. Enables cross-database sources. Kept for backwards compatibility; use `datasharing` for new projects instead. |

Reviewer: ```suggestion
| `ra3_node`  | true | Optional, default `False`. Enables cross-database sources. Kept for backward compatibility; use `datasharing` for new projects instead. |
```

> | `ra3_node`  | true | Optional, default `False`. Enables cross-database sources. Acts as an alias for `datasharing`; setting `ra3_node: true` enables `SVV_*` metadata queries. Kept for backwards compatibility; use `datasharing` for new projects instead. |

Reviewer 2: We removed tha alias in the implementation to avoid customer issues
```suggestion
| `ra3_node`  | true | Optional, default `False`. Enables cross-database sources. Kept for backwards compatibility; use `datasharing` for new projects instead. |
```

To run certain macros with autocommit, load the profile with autocommit using the `--profile` flag. For more context, please refer to this [PR](https://github.com/dbt-labs/dbt-redshift/pull/475/files).

### `datasharing`

Previously, the Redshift adapter used PostgreSQL-compatible catalog tables (for example, `pg_*`, `information_schema`) for metadata operations such as listing relations, schemas, and columns. These tables only surface objects within the currently connected database, which prevents cross-database operations needed for [Redshift Datasharing](https://docs.aws.amazon.com/redshift/latest/dg/datashare-overview.html).

> Starting `dbt-redshift v1.11.0rc1`, you can set `datasharing: true` in your `profiles.yml` to enable cross-database and cross-cluster access. When enabled, `dbt-redshift` switches metadata queries to Redshift's native `SVV_*` system views. You can then materialize models into a database or cluster other than the one specified in your profile using `{{ config(database='other_db') }}`.

Reviewer 2: ```suggestion
Starting `dbt-redshift v1.11.0rc1`, you can set `datasharing: true` in your `profiles.yml` to enable cross-database and cross-cluster access. When enabled, `dbt-redshift` switches metadata queries to Redshift's native `SHOW` system commands. You can then materialize models into a database or cluster other than the one specified in your profile using `{{ config(database='other_db') }}`.
```

Example configuration:

<File name='profiles.yml'>

```yml
company-name:
  target: dev
  outputs:
    dev:
      type: redshift
      host: hostname.region.redshift.amazonaws.com
      ...
      datasharing: true  # default: false
```

</File>

Once enabled, you can materialize a model into a different database by setting `database` in the model config. For example:

```sql
{{ config(database='other_db') }}

select * from {{ ref('my_model') }}
```

> The following macros switch to `SVV_*` sources when `datasharing: true`:

Reviewer 2: ```suggestion
The following macros switch to `SHOW` commands when `datasharing: true`:
```

| Macro | Without `datasharing` | With `datasharing` |
|-------|-----------------------|-------------------|
| `list_relations_without_caching` | `information_schema.tables` | `SHOW TABLES FROM SCHEMA` |
| `list_schemas`, `check_schema_exists` | `pg_namespace` | `SHOW SCHEMAS FROM DATABASE` |
| `get_columns_in_relation` | `information_schema.columns` | `SHOW COLUMNS FROM TABLE` |

> | Catalog queries | `pg_class`, `pg_tables`, `pg_views` | `SHOW TABLES FROM SCHEMA + svv_redshift_columns` |

Author: ```suggestion
| Catalog queries | `pg_class`, `pg_tables`, `pg_views` | `SHOW TABLES FROM SCHEMA` + `SVV_REDSHIFT_COLUMNS` |
```

| `get_relation_last_modified` | `information_schema.tables` | `SVV_ALL_TABLES` |

> | Materialized view describe | `information_schema.columns` | `SVV_ALL_COLUMNS` |

Reviewer 2: ```suggestion
| `list_relations_without_caching` | `information_schema.tables` | `SHOW TABLES FROM SCHEMA` |
| `list_schemas`, `check_schema_exists` | `pg_namespace` | `SHOW SCHEMAS FROM DATABASE` |
| `get_columns_in_relation` | `information_schema.columns` | `SHOW COLUMNS FROM TABLE` |
| Catalog queries | `pg_class`, `pg_tables`, `pg_views` | `SHOW TABLES FROM SCHEMA + svv_redshift_columns` |
| `get_relation_last_modified` | `information_schema.tables` | `SHOW TABLES FROM SCHEMA` |
| Grants | `pg_user`, `has_table_privilege()` | `SHOW GRANTS ON TABLE` |
```

<br /><br />

> `ra3_node: true` also enables this behavior and is supported for backwards compatibility. For new projects, use `datasharing: true` instead.

Reviewer: ```suggestion
`ra3_node: true` also enables this behavior and is supported for backward compatibility. For new projects, use `datasharing: true` instead.
```

Take note of the following limitations when using `datasharing`:

- Creating views (including materialized views) in another database is not supported.
- Source freshness checks can have a lag of up to 5 minutes.

> - Metadata queries are limited to 10,000 rows. If a database has more than 10,000 schemas, or a schema has more than 10,000 tables, dbt returns metadata errors.

Reviewer 2: ```suggestion
- Metadata queries are limited to 10,000 rows. If a database has more than 10,000 schemas, or a schema has more than 10,000 tables, dbt runs can face unexpected scenarios.
```

Example configuration:

<File name='profiles.yml'>

```yml
company-name:
  target: dev
  outputs:
    dev:
      type: redshift
      host: hostname.region.redshift.amazonaws.com
      ...
      datasharing: true  # default: false
```

</File>

Once enabled, you can materialize a model into a different database by setting `database` in the model config. For example:

```sql
{{ config(database='other_db') }}

select * from {{ ref('my_model') }}
```

The following macros switch to `SHOW` commands when `datasharing: true`:

| Macro | Without `datasharing` | With `datasharing` |
|-------|-----------------------|-------------------|
| `list_relations_without_caching` | `information_schema.tables` | `SHOW TABLES FROM SCHEMA` |
| `list_schemas`, `check_schema_exists` | `pg_namespace` | `SHOW SCHEMAS FROM DATABASE` |
| `get_columns_in_relation` | `information_schema.columns` | `SHOW COLUMNS FROM TABLE` |
| Catalog queries | `pg_class`, `pg_tables`, `pg_views` | `SHOW TABLES FROM SCHEMA` + `SVV_REDSHIFT_COLUMNS` |
| `get_relation_last_modified` | `information_schema.tables` | `SHOW TABLES FROM SCHEMA` |
| Grants | `pg_user`, `has_table_privilege()` | `SHOW GRANTS ON TABLE` |

> 

Author: ```suggestion
```
